src/rust/aggregator/src/pipelines/user_history.rs

```diff
@@ -121,6 +126,17 @@ impl Pipeline for UserHistory {
             .fetch_all(&mut transaction as &mut PgConnection)
             .await
             .map_err(|e| PipelineError::ProcessingError(anyhow!(e)))?;
+
+            // Update the batch size based on how many events the current batch contained.
+            let n_events = fill_events.len() + change_events.len();
+            if n_events > TARGET_EVENTS && self.batch_size > BigDecimal::from(1) {
+                self.batch_size = &self.batch_size / BigDecimal::from(10);
+                tracing::debug!("Batch size reduced to {}", self.batch_size);
+            } else if n_events < TARGET_EVENTS / 10 && self.batch_size < BigDecimal::from(MAX_BATCH_SIZE) {
+                self.batch_size = &self.batch_size * BigDecimal::from(10);
+                tracing::debug!("Batch size increased to {}", self.batch_size);
+            }
```

Comment:
- [ ] Abstract into util function like `check_batch_size()` that is used across modules
- [ ] Consider using a trait for the two pipelines